### PR TITLE
[improve][pip] PIP-303: Add optional parameters for getPartitionedStats

### PR DESCRIPTION
**Pull request by crossoverJie**
Releted issue: https://github.com/apache/pulsar/issues/21200

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Review by BewareMyPower — COMMENTED** [review]
More overloads will be added in this way. I'm wondering if we can just pass a `XXXOption` to replace so many parameters?

**Comment by crossoverJie** [comment]
> More overloads will be added in this way. I'm wondering if we can just pass a `XXXOption` to replace so many parameters?

Good advice, I'll use `GetPartitionedStatsOptions` instead of `exclusivePublishers/exclusiveSubscriptions`.

```java
@Data
@Builder
public class GetPartitionedStatsOptions {
    /**
     * Set to true to get precise backlog, Otherwise get imprecise backlog.
     */
    private final boolean getPreciseBacklog;

    /**
     * Whether to get backlog size for each subscription.
     */
    private final boolean subscriptionBacklogSize;

    /**
     * Whether to get the earliest time in backlog.
     */
    private final boolean getEarliestTimeInBacklog;
}

CompletableFuture<PartitionedTopicStats> getPartitionedStatsAsync(
        String topic, boolean perPartition, boolean getPreciseBacklog, boolean subscriptionBacklogSize,
        boolean getEarliestTimeInBacklog, GetPartitionedStatsOptions getPartitionedStatsOptions);
```
Any other suggestions?

**Comment by BewareMyPower** [comment]
I think we can package all parameters other than `topic` and `perPartition`, which are required. You can see the following overload:

```java
    default PartitionedTopicStats getPartitionedStats(String topic, boolean perPartition) throws PulsarAdminException {
        return getPartitionedStats(topic, perPartition, false, false, false);
    }
```

**Comment by crossoverJie** [comment]
Great, I'll create a class `GetPartitionedStatsOptions`:
```java
@Data
@Builder
public class GetPartitionedStatsOptions {
    /**
     * Set to true to get precise backlog, Otherwise get imprecise backlog.
     */
    private final boolean getPreciseBacklog;

    /**
     * Whether to get backlog size for each subscription.
     */
    private final boolean subscriptionBacklogSize;

    /**
     * Whether to get the earliest time in backlog.
     */
    private final boolean getEarliestTimeInBacklog;

    /**
     * Whether to exclusive publishers.
     */
    private final boolean exclusivePublishers;

    /**
     * Whether to exclusive subscriptions.
     */
    private final boolean exclusiveSubscriptions;
}
```
---
Refer to this API:
https://github.com/apache/pulsar/blob/82237d3684fe506bcb6426b3b23f413422e6e4fb/pulsar-client-admin-api/src/main/java/org/apache/pulsar/client/admin/GetStatsOptions.java#L26-L41
https://github.com/apache/pulsar/blob/c9fba5f1f7edc6c22219d27af1b7a63ab2d17c44/pulsar-client-admin-api/src/main/java/org/apache/pulsar/client/admin/Topics.java#L1136-L1144

If there are no other issues, I'll update the PIP docuement.

**Comment by crossoverJie** [comment]
@BewareMyPower I have updated the PIP, please take a look.

File: pip/pip-303.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        boolean getEarliestTimeInBacklog,
+        boolean exclusivePublishers, boolean exclusiveConsumers){
+        TopicName tn=validateTopic(topic);
+        WebTarget path=topicPath(tn,"stats")
+        .queryParam("getPreciseBacklog",getPreciseBacklog)
+        .queryParam("subscriptionBacklogSize",subscriptionBacklogSize)
+        .queryParam("getEarliestTimeInBacklog",getEarliestTimeInBacklog).
+        .queryParam("exclusivePublishers",exclusivePublishers)
+        .queryParam("exclusiveConsumers",exclusiveConsumers);
+}
```
Comment (BewareMyPower) [inline pip/pip-303.md:83]: Could you adjust the indent and add the blanks to make the code more readable?

File: pip/pip-303.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    private final boolean exclusiveConsumers;
+    
+}
+```
+
+Implement the `getPartitionedStatsAsync` and `getStatsAsync` interface for `org.apache.pulsar.client.admin.internal.TopicsImpl`
+```java
+@Override
+public CompletableFuture<PartitionedTopicStats> getPartitionedStatsAsync(String topic,
+        boolean perPartition, boolean getPreciseBacklog, boolean subscriptionBacklogSize,boolean getEarliestTimeInBacklog,
```
Comment (BewareMyPower) [inline pip/pip-303.md:59]: The method signature is wrong.

Comment (crossoverJie) [reply pip/pip-303.md]: Done.

File: pip/pip-303.md (hunk 1 above)
Comment (crossoverJie) [reply pip/pip-303.md]: Thanks for your suggestion, done.

**Review by BewareMyPower — COMMENTED** [review]
Overall LGTM.

I just think the name would better be "excludeXxxx" rather than "exclusiveXxx" because the logic is excluding the publishers or consumers from the stats, not including "exclusive" publishers or consumers.

**Comment by crossoverJie** [comment]
> Overall LGTM.
> 
> I just think the name would better be "excludeXxxx" rather than "exclusiveXxx" because the logic is excluding the publishers or consumers from the stats, not including "exclusive" publishers or consumers.

This was my clerical error, done.

**Review by BewareMyPower — APPROVED** [review]

File: pip/pip-303.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        @QueryParam("excludeConsumers") @DefaultValue("false") boolean excludeConsumers)
+```
+
+
+Add parameters for `org.apache.pulsar.broker.service.Topic`
+```java
+TopicStatsImpl getStats(boolean getPreciseBacklog, boolean subscriptionBacklogSize,
+        boolean getEarliestTimeInBacklog, boolean excludePublishers, boolean excludeConsumers);
+
+CompletableFuture<? extends TopicStatsImpl> asyncGetStats(boolean getPreciseBacklog,
```
[the remark this answers is not in the document]
Comment (crossoverJie) [reply pip/pip-303.md]: Good suggestion, done.

**Pull request merged**